Expose setting to limit the local depth of markers per command buffer.

florian-kuebler opened the pull request:
Adding a lot of Vulkan debug markers can have a significant performance
impact when capturing those.
On the layer, we already have an option to limit the depth of
markers (per command buffer). This change exposes this setting to the
capture settings dialog and transfer the specified value to the layer.

On the UI side, the setting is split into a boolean (checkbox) if
the depth should be limited at all and a text field (only enabled when
the checkbox is clicked and only allowing numeric strings in uint64 range).
Internally, we use uint64::max as to disable limitation of the depth
(as already done in the layer).

Note that the text fields also excepts empty strings (as otherwise
it is inconvenient to change the first digit), but resets those to
0.

Bug: http://b/180388662
Test: Try different settings in the capture options menu and take
 captures with Vulkan layer enabled.

![Screenshot 2021-02-17 at 11 21 05](https://user-images.githubusercontent.com/2725914/108192490-9c453d00-7114-11eb-89b9-5ea69610cd91.png)
![tooltip](https://user-images.githubusercontent.com/2725914/108192496-9ea79700-7114-11eb-9334-ba6a1b66dd52.png)

ronaldfw commented:
Do we know whether the layer is enabled when we show the settings? It would be nice to disable that part when the Vulkan layer is not enabled and have a tooltip that explains the layer needs to be enabled first.

florian-kuebler commented:
> Do we know whether the layer is enabled when we show the settings? It would be nice to disable that part when the Vulkan layer is not enabled and have a tooltip that explains the layer needs to be enabled first.

Unfortunately, not. Davide and I have discussed this already in a couple of other contexts, but came to the conclusion that we should not try to expose this information by the IPC.

dpallotti commented on src/OrbitQt/CaptureOptionsDialog.h:
```diff
@@ -16,6 +17,24 @@
 
 namespace orbit_qt {
 
+namespace internal {
```
I thought we didn't want nested namespaces? Why do you need it? Is this header file supposed to be used by a different module?

florian-kuebler replied:
I was missing the fact that we don't use it. I thought it is internal detail. Removed the namespace, as the header is not exposed.

dpallotti commented on src/OrbitQt/CaptureOptionsDialog.cpp:
```diff
@@ -29,4 +31,32 @@ void CaptureOptionsDialog::SetCollectThreadStates(bool collect_thread_state) {
+  return result;
+}
+
+void CaptureOptionsDialog::resetLocalMarkerDepthLineEdit() {
```
Is "reset" lowercase on purpose?

florian-kuebler replied:
Nope, changed it. Thanks!

dpallotti commented on src/OrbitQt/CaptureOptionsDialog.cpp:
```diff
@@ -29,4 +31,32 @@ void CaptureOptionsDialog::SetCollectThreadStates(bool collect_thread_state) {
+  if (ui_->localMarkerDepthLineEdit->text().isEmpty()) {
+    ui_->localMarkerDepthLineEdit->setText(QString::number(0));
+  }
+};
```
```suggestion
}
```

florian-kuebler replied:
Done.

dpallotti commented on src/OrbitQt/CaptureOptionsDialog.ui:
```diff
@@ -9,36 +9,206 @@
    <string>Capture Options</string>
   </property>
+  <property name="toolTip">
+   <string>Limits the maximal depth of Vulkan debug markers per command buffer to the given value.
```
I would use "maximum".

florian-kuebler replied:
obsolete.

dpallotti commented on src/OrbitQt/CaptureOptionsDialog.ui:
```diff
@@ -9,36 +9,206 @@
    <item row="0" column="0">
     <layout class="QVBoxLayout" name="verticalLayout">
+     <item>
+      <layout class="QHBoxLayout" name="horizontalLayout_3">
```
I would suggest using group boxes rather than lines. The way it's done with line-label-line in a horizontal layout is weird to me.

florian-kuebler replied:
Done.

dpallotti reviewed (commented):
I'll leave the final word on the look and also on the Qt xml stuff I know little about to Florian, but here's my opinion:

- From the screenshot I find the controls waaay too crowded, they would definitely benefit from some margins and some space to breathe.
- I suggest using group boxes.
- The line above "These setting will only apply..." looks redundant to me. Also it appears to have the wrong spacing to the label compared to the line below the label.

dpallotti commented:
Another thought: Do we need two labels? "Limit local marker depth per command buffer" and "Max local marker depth per command buffer" basically say the same thing.
We could have the checkbox say "Limit local marker depth per command buffer to:" followed by the textbox?

ronaldfw commented:
> > Do we know whether the layer is enabled when we show the settings? It would be nice to disable that part when the Vulkan layer is not enabled and have a tooltip that explains the layer needs to be enabled first.
> 
> Unfortunately, not. Davide and I have discussed this already in a couple of other contexts, but came to the conclusion that we should not try to expose this information by the IPC.

Ok, then I would suggest that we at least indicate in the tooltip that these settings only apply when the target process has been started with our Vulkan layer enabled. Right now, as a user seeing these settings, I would assume that the Vulkan layer is auto-enabled, also because the next sentence is "These settings will only apply starting from your next capture". 

We don't have to get into details how to enable it (we have documentation for that), only that the target process has to be started with the layer enabled.

florian-kuebler commented:
> > > Do we know whether the layer is enabled when we show the settings? It would be nice to disable that part when the Vulkan layer is not enabled and have a tooltip that explains the layer needs to be enabled first.
> > 
> > 
> > Unfortunately, not. Davide and I have discussed this already in a couple of other contexts, but came to the conclusion that we should not try to expose this information by the IPC.
> 
> Ok, then I would suggest that we at least indicate in the tooltip that these settings only apply when the target process has been started with our Vulkan layer enabled. Right now, as a user seeing these settings, I would assume that the Vulkan layer is auto-enabled, also because the next sentence is "These settings will only apply starting from your next capture".
> 
> We don't have to get into details how to enable it (we have documentation for that), only that the target process has to be started with the layer enabled.

I added a tooltip for the whole Vulkan layer group.

Also I addressed the review comments so far. Here it how it looks now (on Linux -- will try Windows tomorrow):
![Screenshot 2021-02-17 at 18 11 14](https://user-images.githubusercontent.com/2725914/108240866-8b63ee00-714b-11eb-9484-f480a299c101.png)

reichlfl commented:
> > > > Do we know whether the layer is enabled when we show the settings? It would be nice to disable that part when the Vulkan layer is not enabled and have a tooltip that explains the layer needs to be enabled first.
> > > 
> > > 
> > > Unfortunately, not. Davide and I have discussed this already in a couple of other contexts, but came to the conclusion that we should not try to expose this information by the IPC.
> > 
> > 
> > Ok, then I would suggest that we at least indicate in the tooltip that these settings only apply when the target process has been started with our Vulkan layer enabled. Right now, as a user seeing these settings, I would assume that the Vulkan layer is auto-enabled, also because the next sentence is "These settings will only apply starting from your next capture".
> > We don't have to get into details how to enable it (we have documentation for that), only that the target process has to be started with the layer enabled.
> 
> I added a tooltip for the whole Vulkan layer group.
> 
> Also I addressed the review comments so far. Here it how it looks now (on Linux -- will try Windows tomorrow):
> ![Screenshot 2021-02-17 at 18 11 14](https://user-images.githubusercontent.com/2725914/108240866-8b63ee00-714b-11eb-9484-f480a299c101.png)

"General" and "Vulkan Layer" are the titles of the group boxes, aren't they? It's weird that they are on top of the boxes, but apparently that's how Qt handles it on Linux, so I wouldn't change the default behavior...

Would it be more clear to explicitly add "(requires game execution with --orbit)" to "Vulkan Layer", or is that too much text?

reichlfl commented on src/OrbitQt/CaptureOptionsDialog.ui:
```diff
@@ -9,34 +9,135 @@
        <property name="orientation">
         <enum>Qt::Vertical</enum>
        </property>
+       <property name="sizeHint" stdset="0">
```
Is this a vspacer of 0 height?

florian-kuebler replied:
That was also a leftover (with zero effect). I've set this to default again.

reichlfl reviewed (approved):
I have mainly looked at the XML / Look now. Instead of working with spacers in the UI, maybe take a look at the spacing options in the vertical layout that surrounds the group boxes. I think it still looks a bit "condensed", but way better than the initial version. As I wrote above, usually the default values in Qt layouts look good to me, maybe this is also just due to Linux.

dpallotti commented on src/OrbitQt/CaptureOptionsDialog.ui:
```diff
@@ -9,34 +9,135 @@
+            <property name="orientation">
+             <enum>Qt::Horizontal</enum>
+            </property>
+            <property name="sizeHint" stdset="0">
```
It this a spacer at the end of the line?

florian-kuebler replied:
Oh, yeah I missed to remove this one. Done.

dpallotti commented on src/OrbitQt/CaptureOptionsDialog.ui:
```diff
@@ -9,34 +9,135 @@
+      </widget>
+     </item>
+     <item>
+      <spacer name="verticalSpacer_2">
```
Is this spacer between the group boxes? Can you do it with a margin?

florian-kuebler replied:
Done. Used the layout options of the verticalLayout.

dpallotti commented on src/OrbitQt/CaptureOptionsDialog.ui:
```diff
@@ -9,34 +9,135 @@
+             <set>Qt::ImhDigitsOnly</set>
+            </property>
+            <property name="text">
+             <string>0</string>
```
Is 0 the default value for the max depth? When I first check the checkbox, is 0 the first value that is automatically inserted? I would think it makes more sense to have a positive value that we think is sensible.

florian-kuebler replied:
Changed it to 3 and re-added the tooltip.

dpallotti reviewed (approved):
I agree with @reichlfl and strongly suggest to use margins instead of spacers for fixed-sized spaces between components.

I would also change the default limit from 0, I find it a bit confusing (as it means that checking the checkbox without changing the value is equivalent to disabling markers altogether).

Other than that it looks good.

florian-kuebler commented:
> > > > > Do we know whether the layer is enabled when we show the settings? It would be nice to disable that part when the Vulkan layer is not enabled and have a tooltip that explains the layer needs to be enabled first.
> > > > 
> > > > 
> > > > Unfortunately, not. Davide and I have discussed this already in a couple of other contexts, but came to the conclusion that we should not try to expose this information by the IPC.
> > > 
> > > 
> > > Ok, then I would suggest that we at least indicate in the tooltip that these settings only apply when the target process has been started with our Vulkan layer enabled. Right now, as a user seeing these settings, I would assume that the Vulkan layer is auto-enabled, also because the next sentence is "These settings will only apply starting from your next capture".
> > > We don't have to get into details how to enable it (we have documentation for that), only that the target process has to be started with the layer enabled.
> > 
> > 
> > I added a tooltip for the whole Vulkan layer group.
> > Also I addressed the review comments so far. Here it how it looks now (on Linux -- will try Windows tomorrow):
> > ![Screenshot 2021-02-17 at 18 11 14](https://user-images.githubusercontent.com/2725914/108240866-8b63ee00-714b-11eb-9484-f480a299c101.png)
> 
> "General" and "Vulkan Layer" are the titles of the group boxes, aren't they? It's weird that they are on top of the boxes, but apparently that's how Qt handles it on Linux, so I wouldn't change the default behavior...
> 
> I think the space between the two group boxes is still a bit small, is that a vertical layout with default settings? Usually Qt does a decent job with its default spacings and alignments on Windows.
> 
> Would it be more clear to explicitly add "(requires game execution with --orbit)" to "Vulkan Layer", or is that too much text?

I removed the vertical spacer and adjusted the layout options a bit. Also added a note to the Vulkan layer title, and set the default value to 3.

![Screenshot 2021-02-18 at 10 11 26](https://user-images.githubusercontent.com/2725914/108334287-112a7c80-71d2-11eb-9056-d71c836e94d7.png)

florian-kuebler commented:
Merged this now. Feel free to reach out for follow up improvements.
